### ISSUE TITLE
Using objdump to inspect linkages

### ISSUE THREAD

**Issue by jakirkham** [conda/conda-build#2794]
Recently ran into an issue where [`conda-build` 2 reported a linkage]( https://circleci.com/gh/conda-forge/protobuf-feedstock/208 ) on a dependency, `zlib` that turned out to [not actually be a dependency]( https://github.com/google/protobuf/issues/4344 ). Figured I should retest with `conda-build` 3 and found the issue persisted.

Was looking through the `conda-build` code and found plenty of usages of `ldd`, but no usages of `objdump`. Given `objdump` helped identify this issue for me, was wondering if we could start making use `objdump` to help clarify similar situations for myself and others. Possibly as something under `conda inspect`?

**Comment by jakirkham** [comment]
How should it be used from the CLI?

**Comment by mingwandroid** [comment]
> How should it be used from the CLI?

I don't see any value in `conda inspect objects` and `conda inspect linkages` as things that are selectively called or not from the CLI.

This pyldd stuff is run at the end of every build automatically. It prints out, for example:

```
   INFO (r-base,lib/R/bin/Rscript): Needed DSO /usr/lib/libSystem.B.dylib not found in any CDT/compiler package?!
   INFO (r-base,lib/R/library/methods/libs/methods.so): lib/R/lib/libR.dylib found in this package
   INFO (r-base,lib/R/library/methods/libs/methods.so): Needed DSO /usr/lib/libSystem.B.dylib not found in any CDT/compiler package?!
   INFO (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO lib/libpangocairo-1.0.0.dylib in lib/R/library/grDevices/libs/cairo.so found in https://repo.continuum.io/pkgs/main::pango-1.40.11-h22f5747_0
   INFO (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO lib/libtiff.5.dylib in lib/R/library/grDevices/libs/cairo.so found in https://repo.continuum.io/pkgs/main::libtiff-4.0.9-h0dac147_0
   INFO (r-base,lib/R/library/grDevices/libs/cairo.so): lib/R/lib/libR.dylib found in this package
   INFO (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO lib/libpng16.16.dylib in lib/R/library/grDevices/libs/cairo.so found in https://repo.continuum.io/pkgs/main::libpng-1.6.32-hd1e8b91_4
   INFO (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO lib/libcairo.2.dylib in lib/R/library/grDevices/libs/cairo.so found in https://repo.continuum.io/pkgs/main::cairo-1.14.10-h913ea44_6
  ERROR (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO lib/libgobject-2.0.0.dylib in lib/R/library/grDevices/libs/cairo.so found in https://repo.continuum.io/pkgs/main::glib-2.53.6-h33f6a65_2
  ERROR (r-base,lib/R/library/grDevices/libs/cairo.so): .. but glib not in reqs/run, i.e. it is overlinked (likely) or a missing dependency (less likely)
   INFO (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO lib/libpango-1.0.0.dylib in lib/R/library/grDevices/libs/cairo.so found in https://repo.continuum.io/pkgs/main::pango-1.40.11-h22f5747_0
   INFO (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO lib/libjpeg.9.dylib in lib/R/library/grDevices/libs/cairo.so found in https://repo.continuum.io/pkgs/main::jpeg-9b-he5867d9_2
   INFO (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO /usr/lib/libSystem.B.dylib not found in any CDT/compiler package?!
  ERROR (r-base,lib/R/library/grDevices/libs/cairo.so): Needed DSO lib/libintl.8.dylib in lib/R/library/grDevices/libs/cairo.so found in https://repo.continuum.io/pkgs/main::gettext-0.19.8.1-h15daf44_3
  ERROR (r-base,lib/R/library/grDevices/libs/cairo.so): .. but gettext not in reqs/run, i.e. it is overlinked (likely) or a missing dependency (less likely)
```

The `ERROR`s will appear as `WARNING`s only unless you add ..

```
conda_build:
  error_overlinking: true
```

.. to `~/.condarc`

If you do add this then packages that have broken/missing DSO dependencies are considered broken and moved to the broken folder and the build fails.

For very special cases you can add `build/missing_dso_whitelist` entries but in general this should not be done.

**Comment by jakirkham** [comment]
I agree that we should move away from calling these things in the recipe build if that's what you are saying. Also agree that the build should just always do this. Have been onboard with that idea for a long time. ( https://github.com/conda/conda-build/pull/1461 ) Though what you have done is much better than that proposal.

That said, there is value in being able to inspect linkages outside of a build (e.g. a simple `conda` environment). The main use case being debugging user's installs. Is there a preferred way to do this that uses `pyldd`? I guess we can import things from `pyldd` and run them in a `python` interpreter or via `python -c`, but it would be a little nicer for users and maintainers if there was an easy CLI to call for this stuff.

Yeah, we are going to have to sort that out given the recipes that are using `yum_requirements.txt` currently. My guess is we just go ahead with your model of packaging these things from the system and requiring them as `conda` packages, which should make everyone happier, but we haven't discussed this yet.

**Comment by mingwandroid** [comment]
> The main use case being debugging user's installs. Is there a preferred way to do this that uses pyldd?

I guess `conda inspect linkages` and `conda inspect objects` (I have no clue why both these exist) could be re-implemented using the new stuff in `conda-build`'s `post.py` (which uses `pyldd.py`).

I'm not sure if https://github.com/conda/conda-build/blob/54f9b398c5f918d84a7a3488c2c43b52770ce0b1/conda_build/post.py#L425-L579 should be moved into pyldd.py and pyldd.py made into its own stand-alone project. We've talked about something like this internally before. I think @nehaljwani wanted to use it in more situations than at the end of conda-build. Discussion is very welcome here (and PRs even more so).
